File: 23_1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length < 1)
            return null;

        ListNode result = lists[0];

        for (int i=1; i<lists.length; ++i) {
            result = mergeTwoList(result, lists[i]);
        }

        return result;
    }

    private ListNode mergeTwoList(ListNode l1, ListNode l2) {
        // l1 and l2 are sorted
        ListNode head = new ListNode(0);
        ListNode head_end = head;
        while(l1 != null && l2 != null) {
            ListNode min_node;
            if (l1.val <= l2.val) {
                min_node = l1;
                l1 = l1.next;
            } else {
                min_node = l2;
                l2 = l2.next;
            }
            head_end.next = min_node;
            head_end = head_end.next;
        }

        if (l1 != null)
            head_end.next = l1;
        if (l2 != null)
            head_end.next = l2;

        return head.next;
    }
}

File: 23_1.py
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return None

        dummy = ListNode()
        new_tail = dummy

        h = []
        for i, l in enumerate(lists):
            if l:
                heappush(h, (l.val * 10000 + i, l))

        while len(h) > 1:
            v, p = heappop(h)
            new_tail.next = p
            new_tail = new_tail.next

            if p.next:
                i = v % 10000
                heappush(h, (p.next.val * 10000 + i, p.next))

        if h:
            new_tail.next = h[0][1]
            new_tail = new_tail.next

        return dummy.next
